### PR TITLE
[RFC-0007] Enable Azure OIDC for Azure DevOps Repository

### PR DESCRIPTION
**Pull request by dipti-pai**
- Controller changes to set the provider options in git authOptions to fetch provider credentials while building git config if `.spec.provider` field is set to `azure`.

- API docs for setting up Azure workload identity in IAC

- Unit tests for testing provider settings in git  config

### CODE REVIEW

File: docs/spec/v1beta2/imageupdateautomations.md [hunk 1]
```diff
@@ -192,6 +192,46 @@ tuned to adjust the Git operation timeout.
 The proxy configurations are also derived from the referenced GitRepository
 source. `GitRepository.spec.proxySecretRef` can be used to configure proxy use.
 
+`GitRepository` can be configured to specify an OIDC
+[provider](https://github.com/dipti-pai/source-controller/blob/git-azure-oidc-auth/docs/spec/v1/gitrepositories.md#provider)
```
Comment (dipti-pai) [inline docs/spec/v1beta2/imageupdateautomations.md:196]: TODO: replace the doc link here and in pre-requisites from fluxcd/source-controller once PR https://github.com/fluxcd/source-controller/pull/1591 is merged to SC repo.

File: go.mod [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module github.com/fluxcd/image-automation-controller
 
-go 1.22.0
+go 1.22.4
```
Comment (stefanprodan) [inline go.mod:3]: Please bump this to 1.22.5 to match source-controller

File: go.mod [hunk 3]
```diff
@@ -153,3 +160,5 @@ require (
 	sigs.k8s.io/structured-merge-diff/v4 v4.4.1 // indirect
 	sigs.k8s.io/yaml v1.4.0 // indirect
 )
+
+replace github.com/fluxcd/source-controller/api => github.com/dipti-pai/source-controller/api v0.0.0-20240913204118-48417bdc8bd7
```
Comment (stefanprodan) [inline go.mod:164]: You can remove this and use the main branch from SC.

Comment (dipti-pai) [reply go.mod:164]: Do we need to do a release of the SC branch from fluxcd:main ? If I remove the replace, source-controller api version v1.3.0 from the May release is being referred which does not have the provider field. Thanks.

Comment (stefanprodan) [reply go.mod:164]: for now you can do `go get github.com/fluxcd/source-controller/api@main`

Comment (dipti-pai) [reply go.mod:164]: Hmm, I tried that however make is failing as its unable to find the file with the new CRD. If I manually try to get the file, I get a "404: Not Found". Am I missing something else ? 

```
curl -s --fail https://raw.githubusercontent.com/fluxcd/source-controller/v1.3.1-0.20240916104646-ccdd523e06d1/config/crd/bases/source.toolkit.fluxcd.io_gitrepositories.yaml \
        -o cache/gitrepositories_v1.3.1-0.20240916104646-ccdd523e06d1.yaml
make: *** [Makefile:98: cache/gitrepositories_v1.3.1-0.20240916104646-ccdd523e06d1.yaml] Error 22
```

File: docs/spec/v1beta2/imageupdateautomations.md [hunk 4]
```diff
@@ -192,6 +192,71 @@ tuned to adjust the Git operation timeout.
 The proxy configurations are also derived from the referenced GitRepository
 source. `GitRepository.spec.proxySecretRef` can be used to configure proxy use.
 
+#### GitRepository Provider
+
+`GitRepository` can be configured to specify an OIDC
+[provider](https://github.com/fluxcd/source-controller/blob/main/docs/spec/v1/gitrepositories.md#provider)
```
[the remark this answers is not in the document]
Comment (dipti-pai) [reply docs/spec/v1beta2/imageupdateautomations.md:198]: Thanks, this makes sense. What is the process to publish the docs to the website? Is it part of the SC release?

File: internal/source/git.go [hunk 5]
```diff
@@ -180,6 +181,17 @@ func getAuthOpts(ctx context.Context, c client.Client, repo *sourcev1.GitReposit
 		return nil, fmt.Errorf("failed to configure authentication options: %w", err)
 	}
 
+	if repo.Spec.Provider != "" && repo.Spec.Provider != sourcev1.GitProviderGeneric {
+		if repo.Spec.Provider == sourcev1.GitProviderAzure {
```
Comment (makkes) [inline internal/source/git.go:185]:
```suggestion
		if repo.Spec.Provider == sourcev1.GitProviderAzure {
```

**Review by stefanprodan — APPROVED** [review]
LGTM

Thanks @dipti-pai 🏅

**Pull request merged**